marimo/_ast/compiler.py

```diff
@@ -91,17 +97,22 @@ def compile_cell(code: str, cell_id: CellId_t) -> CellImpl:
     else:
         expr = "None"
 
-    # store the cell's code in Python's linecache so debuggers can find it
-    body_filename = get_filename(cell_id)
-    last_expr_filename = get_filename(cell_id, suffix="_output")
-    # cache the entire cell's code
-    cache(body_filename, code)
-    if sys.version_info >= (3, 9):
-        # ast.unparse only available >= 3.9
-        cache(
-            last_expr_filename,
-            ast.unparse(expr) if not isinstance(expr, str) else "None",
-        )
+    if is_script:
+        ast.increment_lineno(module, start_line - 1)
+        body_filename = filename
+        last_expr_filename = filename
+    else:
+        # store the cell's code in Python's linecache so debuggers can find it
+        body_filename = get_filename(cell_id)
+        last_expr_filename = get_filename(cell_id, suffix="_output")
+        # cache the entire cell's code
+        cache(body_filename, code)
+        if sys.version_info >= (3, 9):
+            # ast.unparse only available >= 3.9
+            cache(
+                last_expr_filename,
+                ast.unparse(expr) if not isinstance(expr, str) else "None",
+            )
```

Comment:
So there's actually a current bug where `_output` doesn't display a line number because it's not put in cache:

![image](https://github.com/user-attachments/assets/616084a7-75a2-49ed-8289-57e8b0c15554)

https://marimo.app/l/64pjvh

 @akshayka is there a reason it has a different file? what happens if body_filename == last_expr_filename?

Reply:
This is interesting. I added another test and script, `tests/_runtime/script_data/script_exception_with_output.py`, and it isn't picking up the original source filename.

```
Traceback (most recent call last):
  File "/marimo/tests/_runtime/script_data/script_exception_with_output.py", line 15, in <module>
    app.run()
  File "/marimo/marimo/_ast/app.py", line 282, in run
    outputs, glbls = AppScriptRunner(InternalApp(self)).run()
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/app/script_runner.py", line 82, in run
    outputs, defs = self._run_synchronous(
                    ^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/app/script_runner.py", line 111, in _run_synchronous
    output = execute_cell(cell, glbls, self.app.graph)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/executor.py", line 83, in execute_cell
    return EXECUTION_TYPES[execution_type].execute_cell(cell, glbls, graph)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/marimo/marimo/_runtime/executor.py", line 138, in execute_cell
    return eval(cell.last_expr, glbls)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/marimo_223245/__marimo__cell_Hbol__output.py", line 2, in <module>
NameError: name 'y' is not defined
```

Looking at the traceback, this script runs through `eval(cell.last_expr, glbls)` whereas `tests/_runtime/script_data/script_exception.py` runs through `exec(cell.body, glbls)`. I think the `cell.last_expr.body` attributes need to be updated for the new lineno and col_offset as well to point to the original script. I looked at their values in the debugger and they seemed to correspond to generated files.

I still don't understand the compile_cell() block of code well, but I can come back to this to try a couple of things. It might take me some time to figure this out:
- See if last_expr.body needs to be updated
- See if the original script code can be cached along with the new filenames